crates/ruff/src/rules/pycodestyle/rules/logical_lines/mod.rs

```diff
@@ -100,9 +101,25 @@ impl<'a> LogicalLines<'a> {
                 TokenKind::Rbrace | TokenKind::Rpar | TokenKind::Rsqb => {
                     parens -= 1;
                 }
-                TokenKind::Newline | TokenKind::NonLogicalNewline | TokenKind::Comment
-                    if parens == 0 =>
-                {
+                TokenKind::Comment if parens == 0 => {
+                    // If a comment is followed by a newline, ignore it, and we'll build the line
+                    // when we process the newline. Otherwise, we'll end up creating one logical
+                    // line here, and then another, empty logical line when we process the newline.
+                    //
+                    // The lexer will always emit a newline after a comment _unless_ the comment
+                    // appears at the start of a logical line.
+                    if let Some((token, ..)) = iter.peek() {
+                        let token_kind = TokenKind::from_token(token);
+                        if matches!(
+                            token_kind,
+                            TokenKind::Newline | TokenKind::NonLogicalNewline
+                        ) {
+                            continue;
+                        }
+                    }
```

Comment:
Alternatively, we may be able to get away with checking if this is the first token in the line? In theory that should also work. It's arguably more reliant on details of the lexer, but it's also making the expectation more explicit.

Reply:
Should we fix this in the lexer instead? It feels odd that the newline is conditional on the comment placement. I would expect that a comment is always followed by a newline.

Nit. You can match directly on the Tok. Using TokenKind has some overhead that is only worth it when there are many reads.

Reply:
Honestly not sure. \cc @youknowone - do you know, if it's intentional that we don't emit a newline when a comment is at start-of-line?

Reply:
It may not actually be a bug, I'd have to compare to CPython.

Reply:
It looks like CPython _does_ emit a non-logical newline, even for entirely empty lines, unlike RustPython.

RustPython also has this test:

```rust
#[test]
fn test_logical_newline_line_comment() {
    let source = "#Hello\n#World";
    let tokens = lex_source(source);
    assert_eq!(
        tokens,
        vec![
            Tok::Comment("#Hello".to_owned()),
            // tokenize.py does put an NL here...
            Tok::Comment("#World".to_owned()),
            // ... and here, but doesn't seem very useful.
        ]
    );
}
```

@youknowone - Do you know if this is intentional, or should I fix it?

Reply:
Put up https://github.com/RustPython/Parser/pull/27. (We'll still need code changes here to fix this bug, once that merges.)

Reply:
Thank you! Makes sense. Is the change to remove the explicit `Comment` handling and only rely on the presence of `Newline` and `NonLogicalNewline`?

Reply:
Thank you! I don't think RustPython intended to have different behavior to CPython.
I will look in the patch.